Google calendar integration

toots opened the pull request:
This is a redo of https://github.com/NOLA-musicians-4-Palestine/nm4p_website/pull/2 with the latest changes merged.

beatboxchad committed:
add Containerfile.dev

beatboxchad committed:
install some Python dependencies for Google API in dev container

beatboxchad committed:
a messy first draft that runs on dev machine

beatboxchad committed:
update container build

beatboxchad committed:
Rename Containerfile

beatboxchad committed:
add default pages deploy job

beatboxchad committed:
python script that generates token, auth needs fleshing out

beatboxchad committed:
remove CNAME

beatboxchad committed:
Containerfile extensions are silly

beatboxchad committed:
slugify downloaded flier attachment filenames

get their extension from the 'mimeType' property in Google API
response, slugify everything else. Some minor refactoring included,
tidying var name+scope

beatboxchad committed:
improve stdout logging in tools/sync_calendars.py

beatboxchad committed:
generated post's event description defaults to API event summary

beatboxchad committed:
first attempt to run CI

beatboxchad committed:
second attempt at CI

beatboxchad committed:
fix typo in pages deploy job

beatboxchad committed:
fiddling to get the deploy job to run in container

beatboxchad committed:
whitespace-cleanup

beatboxchad committed:
untabify

beatboxchad committed:
lettuce try it this way

beatboxchad committed:
this should fix the assets

beatboxchad committed:
calendar sync job depends on tools container

beatboxchad committed:
make it deploy even tho auth isn't configured

beatboxchad committed:
I put that needs: clause in the wrong place

beatboxchad committed:
who needs dependencies anyhow

beatboxchad committed:
use relative urls for media and links

beatboxchad committed:
sync calendars with auth token configured

yee, and I cannot stress this enough, haw

beatboxchad committed:
refactor Google API service provisioning to mainer functions

beatboxchad committed:
autoformat (black) sync_calendars.py

beatboxchad committed:
correct workdir of sync calendar CI build step

beatboxchad committed:
use the correct environment for oauth token injection

beatboxchad committed:
switch to liquid template processing of event date

beatboxchad committed:
fix flier path rendering

beatboxchad committed:
refresh auth token before event sync if needed

beatboxchad committed:
directly inject oauth credentials from secret

beatboxchad committed:
run the pages build and deploy hourly on the live branch

beatboxchad committed:
format tools/generate_token.py with black

toots committed:
Merge remote-tracking branch 'origin/live' into google_calendar_integration

toots committed:
Merge remote-tracking branch 'origin/live' into google_calendar_integration

Simonmoush commented on .github/workflows/build-tools-container.yml:
```diff
@@ -0,0 +1,41 @@
+          echo "${{ secrets.GITHUB_TOKEN }}" | docker login ghcr.io -u ${{ github.actor }} --password-stdin
+
+      # Step 3: Try pulling the existing image (if it exists)
+      - name: Pull existing image (if available)
+        continue-on-error: true  # Continue even if the image doesn't exist
+        run: |
+          docker pull ghcr.io/${{ github.repository_owner }}/${{ github.repository }}/tools:${{ github.ref_name }} || echo "No existing image found"
+
+      # Step 4: Build Docker image (will use cached layers if pull was successful)
+      - name: Build Docker image
+        run: |
+          docker build -f tools/Containerfile \
+            -t ghcr.io/${{ github.repository_owner }}/${{ github.repository }}/tools:${{ github.ref_name }} tools
+
+      # Step 5: Push the image to GHCR
+      - name: Push Docker image to GHCR
+        run: |
+          docker push ghcr.io/${{ github.repository_owner }}/${{ github.repository }}/tools:${{ github.ref_name }}
```
The checker says that the tag name cannot be upper case. I'm thinking we just transform the repo names to lower case for the tags.

toots replied:
I'd suggest to confirm that with an actual docker push. I wouldn't be surprised if that worked.

Otherwise, it looks like `github` actions' [limited functions](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions) does not have one for converting the string to lowercase.

In this case, you'd need to convert the name to lowercase in. previous step, output the result and use it in this step.

Simonmoush replied:
I'm not sure if this effectively checked, but I pushed to the branch google_calendar_integration and it did fail saying the name must be lower case.

Simonmoush replied:
Ok this builds now. Pushes and pulls!

Simonmoush commented on .github/workflows/pages-deploy.yml:
```diff
@@ -1,49 +1,68 @@
-  # on pushes to the `live` branch
-  push:
-    branches:
-      - live
```
Want to keep this actually

Simonmoush commented on .github/workflows/pages-deploy.yml:
```diff
@@ -1,49 +1,68 @@
       - name: Checkout
         uses: actions/checkout@v4
+
       - name: Setup Pages
         id: pages
         uses: actions/configure-pages@v5
-      - name: generate CNAME file if a custom domain is defined for the repository
+
+      - name: Inject OAuth Token
+        run: |
+          echo '${{ secrets.GOOGLE_OAUTH_TOKEN }}' > ${{ github.workspace }}/token.json
+
+      - name: Fetch and Render Events from Google Calendar
         run: |
-          CUSTOM_DOMAIN=$(curl -H "Authorization: Bearer ${{ secrets.GITHUB_TOKEN }}" \
-            https://api.github.com/repos/${{ github.repository }}/pages | jq -r '.cname')
-          if [ "$CUSTOM_DOMAIN" != "null" ]; then
-            echo "$CUSTOM_DOMAIN" | tee CNAME
-          fi
+          docker run --rm \
+          -w /workspace \
+          -v ${{ github.workspace }}:/workspace \
+          ghcr.io/${{ github.repository_owner }}/${{ github.repository }}/tools:${{ github.ref_name }} \
+          python tools/sync_calendars.py
+
```
want to keep all this

Simonmoush commented on tools/generate_token.py:
How does this file get run?

Simonmoush reviewed (changes requested):
These are the changes I'm gonna make before merging. Github seems to encourage me to make my comments a review so i did that, don't know if it's appropriate

Simonmoush committed:
Merge branch 'live' into google_calendar_integration

Simonmoush committed:
change github container registry names to lowercase

Simonmoush committed:
add permission to write packages to GITHUB-TOKEN

Simonmoush committed:
update lowercase step to latest form

Simonmoush committed:
keep most stuff from sammy's original work

Simonmoush committed:
build but don't deploy from all branches for testing

Simonmoush committed:
try this...

Simonmoush committed:
comment out a bunch of things I don't need to test if this works

Simonmoush committed:
fix lowercase container registry requirement here as well

Simonmoush committed:
generate creds on google cloud services and use that in secrets

Simonmoush committed:
print the secret google token

Simonmoush committed:
wc secret

Simonmoush committed:
remove newline from token

Simonmoush committed:
change the python file that pages-deploy runs to test

Simonmoush committed:
authenticate to ghcr in deploy step

Simonmoush committed:
try with the sync_calendars.py script

Simonmoush committed:
changed the way some shit is commented... nothing

Simonmoush committed:
WIP: modify sync code to not get infinite events

Simonmoush committed:
let it configure pages, and deploy and upload the artifact

Simonmoush committed:
fix indentation error

Simonmoush committed:
identified a problem with using the refresh token

Simonmoush committed:
Merge branch 'google_calendar_integration' of github.com:NOLA-musicians-4-Palestine/nm4p_website into google_calendar_integration

Simonmoush committed:
Merge branch 'live' into google_calendar_integration

Simonmoush committed:
ignore token and credentials, those shouldn't be published

Simonmoush committed:
add a few params in generate_token

Simonmoush committed:
test service account authentication

Simonmoush committed:
no module google?

I got an error on the first line of my code saying module google not
found, which is strange because it should be running on the same image
that the old file is running on. and the problem doesn't show up there
does it? let's find out

Simonmoush committed:
try with the same imports from sync_calendars.py

maybe importing one of the previous things allows the google module to
be loaded?!?!

Simonmoush committed:
make sure the script is being run in the container

maybe newlines and comments fuck with it

Simonmoush committed:
try unnamed scopes

Simonmoush committed:
remove space from json file in secret

Simonmoush committed:
remove comment

Simonmoush committed:
change auth from oauth to service account

Simonmoush committed:
troubleshooting getting no events

Simonmoush committed:
print events the old way

Simonmoush committed:
oops

Simonmoush committed:
comment out some stuff for sanity check

Simonmoush committed:
get events for specific calendar only

Simonmoush committed:
publish from this branch

Simonmoush committed:
add google drive scope

Simonmoush committed:
remove template text, add location to events

Simonmoush committed:
Only display information that we'll actually have

orgOrBandName was great when we were inputting that information manually
but now that it's pulling from the google calendar there's no good way
to encode that information, so just don't show it.

also remove it from the template along with some other stuff we didn't
really need like the links (i checked that links in the google calendar
event description carry through to our page anyways)

Simonmoush committed:
try to get the right day

Simonmoush committed:
try this

Simonmoush committed:
sometimes the timezone doesn't exist

Simonmoush committed:
test for correct names

Simonmoush committed:
fix day and try another timezone fix

Simonmoush committed:
oops

Simonmoush committed:
oops again

Simonmoush committed:
check if location exists before accessing

Simonmoush committed:
oops yet again

Simonmoush committed:
try pagination, and a few other things

Simonmoush committed:
time bounds on events fetching

Simonmoush committed:
explicitly indicate datetime module

Simonmoush committed:
month is not a real timedelta

Simonmoush committed:
revamp date stuff

Simonmoush committed:
catch some nulls

Simonmoush committed:
investigate

Simonmoush committed:
space head

Simonmoush committed:
just print the start to see what it is

Simonmoush committed:
more investigation

Simonmoush committed:
thorough investigation...

Simonmoush committed:
forgot colon

Simonmoush committed:
dateTime

Simonmoush committed:
save my place

Simonmoush committed:
finally google calendar integration works
